File: app/src/main/java/dev/chu/architecture/java/p87/MainActivityModule.java
package dev.chu.architecture.java.p87;

import javax.inject.Named;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.android.AndroidInjector;
import dagger.android.ContributesAndroidInjector;
import dagger.multibindings.ClassKey;
import dagger.multibindings.IntoMap;
import dev.chu.architecture.etc.ActivityScope;
import dev.chu.architecture.etc.FragmentScope;

//@Module(subcomponents = MainFragmentComponent.class)
//public class MainActivityModule {
//
//    @Provides
//    @ActivityScope
//    public String provideActivityName() {
//        return MainActivity.class.getSimpleName();
//    }
//
//}

// 1. 보일러 플레이트 코드 제거
// MainFragment에 멤버 인젝션을 하기 위해 MainFragmentSubcomponent를 연결

//@Module(subcomponents = MainFragmentSubcomponent.class)
//public abstract class MainActivityModule {
//
//    // @ActivityScope 에서 문자열을 제공하는 프로바이드 메서드
//    @Named("activity")
//    @Provides
//    @ActivityScope
//    static String provideString() {
//        return "String from MainActivityModule";
//    }
//
//    // MainFragment를 위한 인젝터 팩토리를 멀티 바인딩한다.
//    @Binds
//    @IntoMap
//    @ClassKey(MainFragment.class)
//    abstract AndroidInjector.Factory<?> bindInjectorFactory(MainFragmentSubcomponent.Factory factory);
//
//}

// 2. ContributesAndroidInjector 사용

@Module
public abstract class MainActivityModule {

    @Named("activity")
    @Provides
    @ActivityScope
    static String provideString() {
        return "String from MainActivityModule";
    }

    @FragmentScope
    @ContributesAndroidInjector(modules = MainFragmentModule.class)
    abstract MainFragment mainFragment();
}

File: app/src/main/java/dev/chu/architecture/kotlin/p87/MainActivityModule.kt
package dev.chu.architecture.kotlin.p87

import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import dev.chu.architecture.di.ActivityScope
import dev.chu.architecture.di.FragmentScope
import javax.inject.Named

//@Module(subcomponents = [MainFragmentComponent::class])
//class MainActivityModule {
//    @Provides
//    @ActivityScope
//    fun provideActivityName(): String = MainActivity::class.java.simpleName
//}

// 2. 보일러 플레이트 코드 제거

//@Module(subcomponents = [MainFragmentSubcomponent::class])
//abstract class MainActivityModule {
//    companion object {
//        @Named("activity")
//        @Provides
//        @ActivityScope
//        fun provideString(): String = "String from MainActivityModule"
//    }
//
//    @Binds
//    @IntoMap
//    @ClassKey(MainFragment::class)
//    abstract fun bindInjectorFactory(factory: MainFragmentSubcomponent.Factory): AndroidInjector.Factory<*>
//}

// 3. @ContributesAndroidInjector 애노테이션 활용

@Module
abstract class MainActivityModule {
    companion object {
        @Named("activity")
        @Provides
        @ActivityScope
        fun provideString(): String = "String from MainActivityModule"
    }

    @FragmentScope
    @ContributesAndroidInjector(modules = [MainFragmentModule::class])
    abstract fun mainFragment(): MainFragment
}